lib/Dialect/ArithExt/IR/ArithExtOps.td

```diff
@@ -0,0 +1,48 @@
+#ifndef LIB_DIALECT_ARITHEXT_IR_ARITHEXTOPS_TD_
+#define LIB_DIALECT_ARITHEXT_IR_ARITHEXTOPS_TD_
+
+include "lib/Dialect/ArithExt/IR/ArithExtDialect.td"
+include "mlir/IR/BuiltinAttributes.td"
+include "mlir/IR/CommonTypeConstraints.td"
+include "mlir/IR/OpBase.td"
+include "mlir/Interfaces/InferTypeOpInterface.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+
+
+class ArithExt_Op<string mnemonic, list<Trait> traits = [Pure]> :
+        Op<ArithExt_Dialect, mnemonic, traits> {
+  let cppNamespace = "::mlir::heir::arith_ext";
+}
+
+def ArithExt_BarrettReduceOp : ArithExt_Op<"barrett_reduce", [SameOperandsAndResultType]> {
+  let summary = "Compute the first step of the Barrett reduction.";
+  let description = [{
+    Let $q$ denote a statically known modulus and $b = 4^{w}$, where $w$ is the
+    smallest bit-width that contains the range $[0, q)$. The Barrett reduce
+    operation computes `barret_reduce x = x - floor(x * floor(b / q) / b) * q`.
+
+    Given $0 <= x < q^2$, then this will compute $(x \mod q)$ or $(x \mod q) + p$.
+  }];
+
+  let arguments = (ins
+    SignlessIntegerLike:$input,
+    I64Attr:$modulus
+  );
+  let results = (outs SignlessIntegerLike:$output);
+  let assemblyFormat = "operands attr-dict `:` qualified(type($input))";
+
+  let hasVerifier = 1;
+}
+
+def ArithExt_SubIfGEOp : ArithExt_Op<"subifge", [SameOperandsAndResultType]> {
+  let summary = "Compute (x >= y) ? x - y : x.";
```

Comment:
is this just to make the lowering look nicer (`arith_ext.subifge` vs `arith.subi` + `arith.cmpi` + `arith.select`) or are there optimization/alternative lowering options? (I.e., do some targets have this as a native op?). I'm just curious, I'm not against having it "just" to make the lowering nicer :)

Reply:
The operation is added for the specific HEaan paper optimizations. They show that you can reduce the number of `arith_ext.subifge` operations by relaxing the output ranges.